File: upstream/ray-gauges/packages/reactor-gen/src/instructions/initConfig.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface InitConfigArgs {
  rayRewardDailyEmission: BN
  isoRayAprBps: number
}

export interface InitConfigAccounts {
  payer: PublicKey
  config: PublicKey
  /** Vault to hold RAY tokens that are staked by Reactor depositors */
  rayVault: PublicKey
  /** Hopper to hold RAY rewards that get paid out to Reactor stakers */
  rayHopper: PublicKey
  rayMint: PublicKey
  tokenProgram: PublicKey
  systemProgram: PublicKey
}

export const layout = borsh.struct([
  borsh.u64("rayRewardDailyEmission"),
  borsh.u16("isoRayAprBps"),
])

/** Initialize the reactor config */
export function initConfig(
  args: InitConfigArgs,
  accounts: InitConfigAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.payer, isSigner: true, isWritable: true },
    { pubkey: accounts.config, isSigner: false, isWritable: true },
    { pubkey: accounts.rayVault, isSigner: false, isWritable: true },
    { pubkey: accounts.rayHopper, isSigner: false, isWritable: true },
    { pubkey: accounts.rayMint, isSigner: false, isWritable: false },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([23, 235, 115, 232, 168, 96, 1, 231])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      rayRewardDailyEmission: args.rayRewardDailyEmission,
      isoRayAprBps: args.isoRayAprBps,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}

File: upstream/ray-gauges/sol/programs/reactor/src/instructions/admin/init_config.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{Mint, Token, TokenAccount};

use crate::{
    errors::ReactorErrors, state::ReactorConfig, REACTOR_CONFIG_SEED, REACTOR_REWARD_HOPPER_SEED,
    REACTOR_VAULT_SEED,
};

#[derive(Accounts)]
pub struct InitConfig<'info> {
    #[account(
        mut,
        address = crate::admin::id() @ ReactorErrors::NotAdmin
    )]
    pub payer: Signer<'info>,

    #[account(
        init,
        payer = payer,
        seeds = [
            REACTOR_CONFIG_SEED.as_ref(),
        ],
        bump,
        space = ReactorConfig::LEN
    )]
    pub config: Account<'info, ReactorConfig>,

    /// Vault to hold RAY tokens that are staked by Reactor depositors
    #[account(
        init,
        seeds = [
            REACTOR_VAULT_SEED.as_ref(),
        ],
        bump,
        token::authority = config,
        token::mint = ray_mint,
        payer = payer
    )]
    pub ray_vault: Account<'info, TokenAccount>,

    /// Hopper to hold RAY rewards that get paid out to Reactor stakers
    #[account(
        init,
        seeds = [
            REACTOR_REWARD_HOPPER_SEED.as_ref(),
        ],
        bump,
        token::authority = config,
        token::mint = ray_mint,
        payer = payer
    )]
    pub ray_hopper: Account<'info, TokenAccount>,

    #[account(address = crate::ray_mint::ID)]
    pub ray_mint: Account<'info, Mint>,

    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,
}

pub fn handler(
    ctx: Context<InitConfig>,
    // RAY emitted per day
    ray_reward_daily_emisison: u64,

    // APR for isoRAY in basis points
    iso_ray_apr_bps: u16,
) -> Result<()> {
    let c = &mut ctx.accounts.config;

    c.bump = [ctx.bumps.config];
    c.ray_mint = ctx.accounts.ray_mint.key();
    c.ray_vault = ctx.accounts.ray_vault.key();
    c.rewards_emitted_until = Clock::get()?.unix_timestamp as u64;
    c.ray_reward_hopper = ctx.accounts.ray_hopper.key();
    c.ray_reward_daily_emission = ray_reward_daily_emisison;
    c.iso_ray_apr_bps = iso_ray_apr_bps;

    Ok(())
}
